native/dlaqp2.go

```diff
@@ -0,0 +1,111 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+)
+
+// Dlaqp2 computes a QR factorization with column pivoting of the block A[offset:m, 0:n]
+// of the m×n matrix A. The block A[0:offset, 0:n] is accordingly pivoted, but not factorized.
+//
+// On exit, the upper triangle of block A[offset:m, 0:n] is the triangular factor obtained.
+// The elements in block A[offset:m, 0:n] below the diagonal, together with tau, represent
+// the orthogonal matrix Q as a product of elementary reflectors.
+//
+// offset is number of rows of the matrix A that must be pivoted but not factorized.
+// offset must not be negative otherwise Dlaqp2 will panic.
+//
+// On exit, if jpvt[i] = k, then the ith column of A*P was the kth column of A.
+// jpvt must have length n, otherwise Dlapq2 will panic.
+//
+// On exit tau holds the scalar factors of the elementary reflectors. It must have length
+// at least min(m, n) otherwise Dlapq2 will panic.
+//
+// vn1 and vn2 hold the partial and complete column norms respectively. They must have length n,
+// otherwise Dlapq2 will panic.
+//
+// work must have length m, otherwise Dlapq2 will panic.
+//
+// Dlapq2 is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dlapq2(m, n, offset int, a []float64, lda int, jpvt []int, tau, vn1, vn2, work []float64) {
+	checkMatrix(m, n, a, lda)
+	if len(jpvt) != n {
+		panic(badIpiv)
+	}
+	if len(tau) < min(m, n) {
+		panic(badTau)
+	}
+	if len(vn1) < n {
+		panic(badVn1)
+	}
+	if len(vn2) < n {
+		panic(badVn2)
+	}
+	if len(work) < m {
+		panic(badWork)
+	}
+
+	mn := min(m-offset, n)
+	tol3z := math.Sqrt(dlamchE)
+
+	bi := blas64.Implementation()
+
+	// Compute factorization.
+	for i := 0; i < mn; i++ {
+		offpi := offset + i
+
+		// Determine ith pivot column and swap if necessary.
+		p := i + bi.Idamax(n-i, vn1[i:], 1)
+		if p != i {
+			bi.Dswap(m, a[p:], lda, a[i:], lda)
+			jpvt[p], jpvt[i] = jpvt[i], jpvt[p]
+			vn1[p] = vn1[i]
+			vn2[p] = vn2[i]
+		}
+
+		// Generate elementary reflector H[i].
+		if offpi < m {
```

Comment:
If `offset` is 0 and `m <= n`, then `i == offpi` and it goes from 0 to `m-1`, so it is always less than `m`. The code must handle differently the case when there are no more rows below in the column `i` because then there is nothing to zero out for Dlarfg and `A( OFFPI+1, I )` would index outside of the matrix.

Please recheck but I _think_ that it _should_ be something like:
```
if offpi < m-1 {
    // Note the (offpi+1) in the third argument.
    a[offpi*lda+i], tau[i] = impl.Dlarfg(m-offpi, a[offpi*lda+i], a[(offpi+1)*lda+i:], lda)
} else {
    // Last row, we can call Dlarfg:
    // a[(m-1)*lda+i], tau[i] = impl.Dlarfg(1, a[(m-1)*lda+i], nil, lda)
    // or we can just equivalently do:
    tau[i] = 0
}
```